relay-server/src/services/project/metrics.rs

```diff
@@ -0,0 +1,333 @@
+use std::marker::PhantomData;
+use std::ops::Deref;
+
+use relay_dynamic_config::{ErrorBoundary, Feature, Metrics};
+use relay_filter::FilterStatKey;
+use relay_metrics::aggregator::AggregatorBuckets;
+use relay_metrics::{Bucket, MetricNamespace};
+use relay_quotas::Scoping;
+use relay_system::Addr;
+
+use crate::metric_stats::MetricStats;
+use crate::services::outcome::{Outcome, TrackOutcome};
+use crate::services::project::ProjectState;
+use crate::services::project_cache::BucketSource;
+use crate::utils;
+
+pub struct Filtered;
+pub struct WithProjectState;
+
+/// Container for a vector of buckets.
+#[derive(Debug)]
+pub struct Buckets<State = ()> {
```

Comment:
Does a default for this type parameter even make sense? I would expicitly require an init state instead, e.g. `Buckets<Unfiltered>`.

Reply:
No state is the init state and the default at the same time, you can pass around `Buckets` if you don't care about state until you do. Having something explicit didn't really make it better imo.